OCPEDGE-1483: Add TNF E2E tests for network failure

clobrano opened the pull request:
This commit introduces a new e2e test for TNF to validate etcd recovery after a network disruption in a two-node cluster.

The test simulates a network partition by using iptables to block traffic between the two nodes for a short, configurable, duration.

It then verifies that the cluster correctly handles the split-brain scenario, with one node becoming the leader and the other rejoining as a learner before being promoted back to a voting member.

Additionally, the node disruption pod creation logic has been refactored from `triggerNodeReboot` to a more generic `createNodeDisruptionPod` to accommodate different types of disruptions, including network partitions and reboots.

jaypoulz commented on test/extended/util/nodes.go:
```diff
@@ -95,19 +95,30 @@ func DebugAllNodesRetryWithOptionsAndChroot(oc *CLI, debugNodeNamespace string,
+	return command, triggerNodeDisruption(kubeClient, target.Name, 0, true, command)
+}
+
+func triggerNodeDisruption(kubeClient kubernetes.Interface, nodeName string, attempt int, useHostNetwork bool, command string) error {
```
I don't love the `userHostNetwork` bool.
I would prefer some kind of enum that allows me to specify which kind of disruption.
-> NodeCrash
-> NetworkPartition
etc.

I think this will allow us to reuse the pattern as we add more disruption types (e.g. kubelet failure, etc)

jaypoulz replied:
I get what you mean. Maybe a name tweak would suffice. Instead of "triggerNodeDisruption", could we rename the function to something like: "createNodeDisruptionPod" or "execNodeDisruptionPod"

This would make it clear to that usingHostNetwork is a change that affects the created pod, not something that dictates the kind of disruption. (The kind of disruption is defined by the command / all of the arguments together).

clobrano replied:
Good idea :+1:

jaypoulz commented on test/extended/two_node/tnf_recovery.go:
```diff
@@ -251,7 +299,76 @@ func validateEtcdRecoveryState(e *helpers.EtcdClientFactoryImpl, survivedNode *c
+
+		// Once we get one leader and one learner, we don't care if the latter has started already, but the first must
+		// already been started
+		nonLearnerStarted := (startedA && !learnerA) || (startedB && !learnerB)
```
or leaderStarted

clobrano commented:
Addressed the comments, moreover I changed the disruption command to something more specific that limits _only_ the traffic between the two nodes and not the entire interface.

jaypoulz commented on test/extended/util/nodes.go:
```diff
@@ -95,19 +105,43 @@ func DebugAllNodesRetryWithOptionsAndChroot(oc *CLI, debugNodeNamespace string,
+	sleepCmd := fmt.Sprintf("sleep %d", int(disruptionDuration.Seconds()))
+	disruptionCmd := fmt.Sprintf("%s 'trap \"%s\" EXIT; %s ; %s'", preambleCmd, cleanupCmd, blockTrafficCmd, sleepCmd)
+
+	return disruptionCmd, createNodeDisruptionPod(kubeClient, target.Name, 0, hostNetwork, disruptionCmd)
```
Very clever!

jaypoulz reviewed (approved):
/lgtm

jaypoulz commented:
/hold

clobrano commented:
/pj-rehearse auto-ack

clobrano commented:
/retest-required

eggfoobar reviewed (commented):
Just some small clarity things, looking good.

clobrano commented on test/extended/util/nodes.go:
```diff
@@ -95,19 +105,43 @@ func DebugAllNodesRetryWithOptionsAndChroot(oc *CLI, debugNodeNamespace string,
 }
 
-func triggerNodeReboot(kubeClient kubernetes.Interface, nodeName string, attempt int, command string) error {
+func createNodeDisruptionPod(kubeClient kubernetes.Interface, nodeName string, attempt int, networkMode hostNetworkMode, command string) error {
```
the function does create the node disruption pod, so the name seems pretty accurate to me

clobrano replied:
This function creates a pod with the goal to execute a given command, which might disrupt the network or reboot the node, either gracefully or ungracefully.

stbenjam commented:
/hold cancel
/retest-required

jaypoulz commented:
/retest required
